src/security/PeerConnector.cc

```diff
@@ -694,13 +680,31 @@ Security::PeerConnector::certDownloadingDone(SBuf &obj, int downloadStatus)
         return;
     }
 
-    srvBio->holdRead(false);
-    negotiateSsl();
+    // We should have the full certificates chain now, so clear the
+    // fMissingIssuer flag to report any missing issuer certificates.
+    auto verifyData = Ssl::SquidVerifyData::SessionData(session);
+    assert(verifyData);
+    verifyData->clear(Ssl::SquidVerifyData::fMissingIssuer);
+
+    if (!Ssl::PeerCertificatesVerify(session, downloadedCerts)) {
+        noteNegotiationError(SSL_ERROR_SSL, 0, 0);
+        return;
+    }
+
+    resumeNegotiation();
```

Comment:
It makes sense to resume negotiation after the certificates were downloaded and verification was completed. However, resumeNegotiation() is _conditional_ on the negotiation being suspended previously -- the method silently does nothing if resumeNegotiationCall is unset. This raises a question:

Do we implicitly require that the negotiation is suspended at this point?

If we do, then resumeNegotiation() should probably check the "resumeNegotiationCall must be set" invariant instead of silently ignoring its violations.

Otherwise, we probably need to move the "resumeNegotiationCall is set" condition from resumeNegotiation() into here and then handle (at least via a source code comment) the non-obvious case of resumeNegotiationCall being nil.

Reply:
> This raises a question:
> Do we implicitly require that the negotiation is suspended at this point?

Yes

commit 7d749c4dbcb